operator/pkg/controller/karmada/karmada_controller.go

```diff
@@ -0,0 +1,102 @@
+package karmada
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	operatorv1alpha1 "github.com/karmada-io/karmada/operator/pkg/apis/operator/v1alpha1"
+)
+
+const (
+	// ControllerName is the controller name that will be used when reporting events.
+	ControllerName = "karmada-controller"
+
+	// ControllerFinalizerName is the name of the karmada controller finalizer
+	ControllerFinalizerName = "operator.karmada.io/finalizer"
+)
+
+// Controller controls the Karmada resource.
+type Controller struct {
+	client.Client
+
+	EventRecorder record.EventRecorder
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (ctrl *Controller) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	startTime := time.Now()
+	klog.V(4).InfoS("Started syncing karmada", "karmada", req, "startTime", startTime)
+	defer func() {
+		klog.V(4).InfoS("Finished syncing karmada", "karmada", req, "duration", time.Since(startTime))
+	}()
+
+	karmada := &operatorv1alpha1.Karmada{}
+	if err := ctrl.Get(ctx, req.NamespacedName, karmada); err != nil {
+		// The resource may no longer exist, in which case we stop processing.
+		if errors.IsNotFound(err) {
+			klog.V(2).InfoS("Karmada has been deleted", "karmada", req)
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{}, err
+	}
+
+	// Deep-copy otherwise we are mutating our cache.
+	// TODO: Deep-copy only when needed.
+	karmada = karmada.DeepCopy()
```

Comment:
I don't think the variable `karmada` now using cache. It points to an initialized memory by `&operatorv1alpha1.Karmada{}`.